zerver/lib/user_groups.py

```diff
@@ -10,9 +10,14 @@
 from zerver.models import GroupGroupMembership, Realm, UserGroup, UserGroupMembership, UserProfile
 
 
-def access_user_group_by_id(user_group_id: int, user_profile: UserProfile) -> UserGroup:
+def access_user_group_by_id(
+    user_group_id: int, user_profile: UserProfile, *, for_read: bool = False
+) -> UserGroup:
     try:
         user_group = UserGroup.objects.get(id=user_group_id, realm=user_profile.realm)
+        if for_read:
+            # Everyone is allowed to read a user group and check who are its members.
+            return user_group
```

Comment:
I think the long-term plan is for each user group to have a "Group of which users can see membership of this user group". 

I think this detail should probably match the current API-level question of whether guests get the direct membership of user groups via the API today. 

But I suppose in the medium term, we may end up with an independent `can_access_user_group_membership()` independent function called both here and in the `register` code path for fetching all user groups. Which in turn will require some way to ask the bulk question with a single database query, i.e. "Fetch the user groups that this user can access".

Reply:
We currently do not allow guests to access any API endpoint for user groups, even the one that returns all the user groups of the realm and there is no endpoint to get members of a user groups for now. But the guests can get list of all user groups, including the member ids of them, from the `/register` endpoint.

Reply:
Sounds good. I also extended this like so:
```
+        if for_read and not user_profile.is_guest:
+            # Everyone is allowed to read a user group and check who
+            # are its members. Guests should be unable to reach this
+            # code path, since they can't access user groups API
+            # endpoints, but we check for guests here for defense in
+            # depth.
```

I think having that here will be a helpful reminder when we get to changing the groups logic for guests.